### PR TITLE
functionalization bugfix: using owning type when unwrapping tensors

### PR DESCRIPTION
**Pull request by bdhirsh**
Addresses the comment here: https://github.com/pytorch/pytorch/pull/73441#discussion_r853873880

This PR should also fix a bug with `at::cat.out` in functionalization, so I added a test for it.

When the functionalization pass unwraps tensor arguments, we need to make sure the thing that we unwrap into has an owning type - we can't blindly base the type off whatever types come from the operator's schema.

This is basically just a problem for `at::TensorList` (non-owning) - we need to use an owning `std::vector<Tensor>` instead.

@ezyang I wasn't sure how general of a fix to make. `translate()` already knows a bit about "expensive conversions" from non-owning to owning types. I couldn't think of a way to re-use that code without making some larger changes (like teaching `CType`'s about whether they're owning types), so I added a simple fix inside the functionalization codegen. Feedback totally welcome.

Here's part of the old `cat` kernel:
```
    at::Tensor & cat_out_out(c10::DispatchKeySet dispatchKeySet, at::TensorList tensors, int64_t dim, at::Tensor & out) {
      at::TensorList tensors_;
      if (at::functionalization::impl::isFunctionalTensor(tensors)) {
        at::functionalization::impl::sync(tensors);
        tensors_ = at::functionalization::impl::from_functional_tensor(tensors);
      } else {
        tensors_ = tensors;
      }
      ....
    }
```

And here's the new one (the temporary it creates is a vector)
```
    at::Tensor & cat_out_out(c10::DispatchKeySet dispatchKeySet, at::TensorList tensors, int64_t dim, at::Tensor & out) {
      ::std::vector<at::Tensor> tensors_;
      if (at::functionalization::impl::isFunctionalTensor(tensors)) {
        at::functionalization::impl::sync(tensors);
        tensors_ = at::functionalization::impl::from_functional_tensor(tensors);
      } else {
        tensors_ = tensors;
      }
      ...
    }
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527 [prototype] integrate functionalization <> LTC torchscript backend
* #76320 attempt to functionalize ops with mutable positional-only args
* #76126 add native view_copy.out ops, teach codegen about tensorlist out=
* #76319 [poc] try to fix torch.tensor for functionalization
* #76318 fix nested grad(functionalize(f)) transforms
* **#76125 functionalization bugfix: using owning type when unwrapping tensors**

### CODE REVIEW

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 1]
```diff
@@ -341,6 +341,30 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
...
+void commit_update(const Tensor& functional_tensor) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(isFunctionalTensor(functional_tensor));
+  unsafeGetFunctionalWrapper(functional_tensor)->commit_update();
+}
+
+void commit_update(const TensorList functional_tensor) {
+  for (const auto i : c10::irange(functional_tensor.size())) {
+    commit_update(functional_tensor[i]);
+  }
+}
```
Comment (bdhirsh) [inline aten/src/ATen/FunctionalTensorWrapper.cpp:366]: These aren't strictly necessary for this PR, but I need them later (accidentally bundled them in this commit)

**Review by ezyang — APPROVED** [review]
the one off seems fine. keep an eye on it in the future

**Pull request closed without merging**